sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/AliasHelper.scala

```diff
@@ -70,11 +70,17 @@ trait AliasHelper {
   protected def replaceAliasButKeepName(
      expr: NamedExpression,
      aliasMap: AttributeMap[Alias]): NamedExpression = {
-    // Use transformUp to prevent infinite recursion when the replacement expression
-    // redefines the same ExprId,
-    trimNonTopLevelAliases(expr.transformUp {
+    expr match {
+      // We need to keep the `Alias` if we replace a top-level Attribute, so that it's still a
+      // `NamedExpression`. We also need to keep the name of the original Attribute.
       case a: Attribute => aliasMap.get(a).map(_.withName(a.name)).getOrElse(a)
-    }).asInstanceOf[NamedExpression]
+      case _ =>
+        // Use transformUp to prevent infinite recursion when the replacement expression
+        // redefines the same ExprId,
+        expr.transformUp {
+          case a: Attribute => aliasMap.get(a).map(_.child).getOrElse(a)
```

Comment:
This is the key change. We replace the attribute with `Alias.child`, so that we don't need to call `trimNonTopLevelAliases` to remove it later.